### PR TITLE
Fix #6490 add mybatis extension

### LINKED ISSUE #6490: Provide an extension for MyBatis
Hi, 
Considering using JPA would cost a lot of efforts on migration from original Spring/MyBatis projects to Quarkus. I'm trying with [MyBatis-CDI](http://mybatis.org/cdi/injection.html) for rebuilding the project within Quarkus.  But I got these Build failure Errors:

>  ERROR [io.qua.dev.DevModeMain] Failed to start Quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Found 2 deployment problems:
[1] Unsatisfied dependency for type org.unisce.mapper.CourseMapper and qualifiers [@Default]
        - java member: org.unisce.service.impl.CourseServiceImpl#courseMapper
        - declared on CLASS bean [types=[org.unisce.service.CourseService, java.lang.Object, org.unisce.service.CurdService<org.unisce.model.Course>, org.unisce.service.impl.CourseServiceImpl], qualifiers=[@Default, @Any], target=org.unisce.service.impl.CourseServiceImpl]
[2] Unsatisfied dependency for type org.apache.ibatis.session.SqlSession and qualifiers [@Default]
        - java member: org.unisce.service.impl.CourseServiceImpl#sqlSession
        - declared on CLASS bean [types=[org.unisce.service.CourseService, java.lang.Object, org.unisce.service.CurdService<org.unisce.model.Course>, org.unisce.service.impl.CourseServiceImpl], qualifiers=[@Default, @Any], target=org.unisce.service.impl.CourseServiceImpl]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:858)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:220)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:106)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:251)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:941)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
        Suppressed: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.unisce.mapper.CourseMapper and qualifiers [@Default]

I refer to code [examples ](https://github.com/mnesarco/mybatis-cdi-samples) .

My  SqlSessionFactory Producers:
----------------------------------------------------------------

@ApplicationScoped
public class MyBatisProducers {

    @ApplicationScoped
    @Named("course1")
    @Produces
    @SessionFactoryProvider
    public SqlSessionFactory produceFactory() throws IOException {
        String resource = "mybatis-config.xml";
        Reader reader = Resources.getResourceAsReader(resource);
        SqlSessionFactory sqlSessionFactory = new SqlSessionFactoryBuilder().build(reader);
        reader.close();

        return sqlSessionFactory;
    }
}
----------------------------------------------------------------

My Resource/Controller  Class:
----------------------------------------------------------------

`
@Singleton
@Path("/course")
public class CourseController {

	@Inject
	CourseService courseService;

	@GET
	@Path(value = "/list")
	public HttpResult findAll() {
		return HttpResult.ok(courseService.findAll());
	}
}
`
----------------------------------------------------------------

My Service Interface Class:
----------------------------------------------------------------

`
public interface CourseService  extends CurdService<Course> {
        List<Course> findAll();
}
`

----------------------------------------------------------------

My ServiceImpl Class :
----------------------------------------------------------------
`	
@Singleton
public class CourseServiceImpl implements CourseService {

	@Inject
	CourseMapper courseMapper;

	@Inject
	SqlSession sqlSession;

        @Override
	public List<Course> findAll() {

		List<Course> test2 = sqlSession.getMapper(CourseMapper.class).findAll();
		return test2;
	}
}`
----------------------------------------------------------------

I dont know whether it's caused by my misusing CDI annotation or  any MyBatis-CDI Quarkus problems.  Any professionals could help? Thanks a lot.

### PR DESCRIPTION
**Pull request by zhfeng**

### CODE REVIEW

File: extensions/mybatis/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gastaldi) [inline extensions/mybatis/runtime/src/main/resources/META-INF/quarkus-extension.yaml:6]: I don't see a documentation page for this link. You need to add a `mybatis.adoc` in `docs/src/main/asciidoc`

File: extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java
Comment (gastaldi) [inline extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java:15]: No need to explicitly set the value here. The `defaultValue` in the `@ConfigItem` annotation will already set it for you

File: extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java
Comment (gastaldi) [inline extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java:21]: Same as above

File: extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRecorder.java
Comment (gastaldi) [inline extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRecorder.java:48]: Maybe `log.debug` here?

File: extensions/mybatis/deployment/src/main/java/io/quarkus/mybatis/deployment/MyBatisMapperBuildItem.java
Comment (gastaldi) [inline extensions/mybatis/deployment/src/main/java/io/quarkus/mybatis/deployment/MyBatisMapperBuildItem.java:14]: Can you be more specific on what this `DotName` should be?

File: bom/runtime/pom.xml
Comment (gastaldi) [inline bom/runtime/pom.xml:2802]:
```suggestion
                <artifactId>mybatis</artifactId>
		        <version>${mybatis.version}</version>
```

File: extensions/mybatis/runtime/pom.xml
Comment (gastaldi) [inline extensions/mybatis/runtime/pom.xml:21]: `quakrus-core` is resolved transitively from `quarkus-arc`, so it's not really required to be here but that's ok

File: extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java
Comment (gastaldi) [inline extensions/mybatis/runtime/src/main/java/io/quarkus/mybatis/runtime/MyBatisRuntimeConfig.java:32]:
```suggestion
    @ConfigItem(name = "initial-sql")
```

We prefer kebab style :)

**Review by gastaldi — CHANGES_REQUESTED** [review]
It would be nice to have integration tests (including native generation) too

**Comment by zhfeng** [comment]
Thanks @gastaldi - I just made the changes as your suggestions and add the integration tests for the mybatis. The JVM test works fine but the native test looks weird. The native image build is failing with the following messges
```
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:native-image (native-image) @ quarkus-mybatis-integration-test ---
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building native image source jar: /home/zhfeng/work/zhfeng/quarkus/integration-tests/mybatis/target/quarkus-mybatis-integration-test-999-SNAPSHOT-native-image-source-jar/quarkus-mybatis-integration-test-999-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Building native image from /home/zhfeng/work/zhfeng/quarkus/integration-tests/mybatis/target/quarkus-mybatis-integration-test-999-SNAPSHOT-native-image-source-jar/quarkus-mybatis-integration-test-999-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 20.1.0 (Java Version 11.0.7)
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /home/zhfeng/Applications/graalvm-ce-java11-20.1.0/bin/native-image -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar quarkus-mybatis-integration-test-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeAllTimeZones -H:EnableURLProtocols=http,https --enable-all-security-services -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-mybatis-integration-test-999-SNAPSHOT-runner
-H:IncludeAllTimeZones and -H:IncludeTimeZones are now deprecated. Native-image includes all timezonesby default.
[quarkus-mybatis-integration-test-999-SNAPSHOT-runner:674369]    classlist:   6,446.38 ms,  0.94 GB
[quarkus-mybatis-integration-test-999-SNAPSHOT-runner:674369]        (cap):     718.20 ms,  0.94 GB
[quarkus-mybatis-integration-test-999-SNAPSHOT-runner:674369]        setup:   2,509.04 ms,  0.94 GB
14:18:26,050 INFO  [org.jbo.threads] JBoss Threads version 3.1.1.Final
[quarkus-mybatis-integration-test-999-SNAPSHOT-runner:674369]     analysis:  20,234.45 ms,  2.64 GB
2 fatal errors detected:
Fatal error:com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing com.oracle.svm.reflect.Class_getInterfaces0_afd19408bba3599ae1dfbbcd6448ba114ce3d3fc.invoke(java.lang.Object, java.lang.Object[]) 
Parsing context:
	parsing java.lang.reflect.Method.invoke(Method.java:566)
	parsing io.quarkus.arc.impl.Qualifiers.invoke(Qualifiers.java:105)
	parsing io.quarkus.arc.impl.Qualifiers.hasQualifier(Qualifiers.java:66)
	parsing io.quarkus.arc.impl.Qualifiers.hasQualifier(Qualifiers.java:47)
	parsing io.quarkus.arc.impl.Qualifiers.hasQualifiers(Qualifiers.java:39)
	parsing io.quarkus.arc.impl.ArcContainerImpl.matches(ArcContainerImpl.java:689)
	parsing io.quarkus.arc.impl.ArcContainerImpl.getMatchingBeans(ArcContainerImpl.java:585)
	parsing io.quarkus.arc.impl.ArcContainerImpl.getBeans(ArcContainerImpl.java:441)
	parsing io.quarkus.arc.impl.BeanManagerImpl.getBeans(BeanManagerImpl.java:94)
	parsing io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:95)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
	parsing io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	parsing com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:149)
	parsing com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:184)
	parsing com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

	at com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:138)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:340)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:311)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:112)
	at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:228)
	at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:470)
	at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:542)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:530)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.Class.getInterfaces0() is reachable: The declaring class of this element has been substituted, but this element is not present in the substitution class
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.lookup(AnnotationSubstitutionProcessor.java:183)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:128)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:128)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookupAllowUnresolved(AnalysisUniverse.java:397)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:377)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:75)
	at com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess.lookupJavaMethod(UniverseMetaAccess.java:93)
	at com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaMethod(AnalysisMetaAccess.java:66)
	at com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaMethod(AnalysisMetaAccess.java:39)
	at com.oracle.svm.reflect.hosted.ReflectionSubstitutionType$ReflectiveInvokeMethod.buildGraph(ReflectionSubstitutionType.java:511)
	at com.oracle.graal.pointsto.meta.AnalysisMethod.buildGraph(AnalysisMethod.java:333)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:189)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:352)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:322)
	... 13 more
Fatal error:com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing com.oracle.svm.reflect.Class_allPermDomain_6c5e23148388eb3ffe80e2382beb0f2363f3eeab.get(java.lang.Object) 
Parsing context:
	parsing java.lang.reflect.Field.get(Field.java:418)
	parsing java.net.HttpConnectSocketImpl.doTunnel(HttpConnectSocketImpl.java:171)
	parsing java.net.HttpConnectSocketImpl$2.run(HttpConnectSocketImpl.java:151)
	parsing java.net.HttpConnectSocketImpl$2.run(HttpConnectSocketImpl.java:149)
	parsing com.oracle.svm.core.jdk.Target_java_security_AccessController.doPrivileged(SecuritySubstitutions.java:119)
	parsing java.util.ServiceLoader.getConstructor(ServiceLoader.java:667)
	parsing java.util.ServiceLoader.loadProvider(ServiceLoader.java:898)
	parsing java.util.ServiceLoader$LayerLookupIterator.hasNext(ServiceLoader.java:951)
	parsing java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
	parsing javax.enterprise.inject.spi.CDI.findAllProviders(CDI.java:120)
	parsing javax.enterprise.inject.spi.CDI.getCDIProvider(CDI.java:82)
	parsing javax.enterprise.inject.spi.CDI.current(CDI.java:64)
	parsing io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:94)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	parsing io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
	parsing io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	parsing com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:149)
	parsing com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:184)
	parsing com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

	at com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:138)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:340)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:311)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:112)
	at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:228)
	at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:470)
	at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:542)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:530)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported field java.lang.Class.allPermDomain is reachable: The declaring class of this element has been substituted, but this element is not present in the substitution class
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.lookup(AnnotationSubstitutionProcessor.java:140)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:123)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:123)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookupAllowUnresolved(AnalysisUniverse.java:355)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:325)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:75)
	at com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess.lookupJavaField(UniverseMetaAccess.java:98)
	at com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaField(AnalysisMetaAccess.java:71)
	at com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaField(AnalysisMetaAccess.java:39)
	at com.oracle.svm.reflect.hosted.ReflectionSubstitutionType$ReflectiveReadMethod.buildGraph(ReflectionSubstitutionType.java:384)
	at com.oracle.graal.pointsto.meta.AnalysisMethod.buildGraph(AnalysisMethod.java:333)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:189)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:352)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:322)
	... 13 more
Error: Image build request failed with exit status 1

```
I tried to add ```--report-unsupported-elements-at-runtime``` and the native image build passes, but the test fails with the NPE which seems that the mybatis mapper bean can not be injected.

**Comment by zhfeng** [comment]
Hi @gastaldi ,

Finally I got the native integration tests working but it has to add the ```--report-unsupported-elements-at-runtime``` option to build the native image. The MyBatis has a heavy use of the Java reflections and dynamic object proxies, so it needs more investigation. Anyway, we can support the basic SQL mapping to query, insert and delete currently. I think this is a good start for the mybatis extension and also add the ```mybatis.adoc``` in the docs. I hope it is helpful.

File: docs/src/main/asciidoc/mybatis.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/mybatis.adoc:128]:
```suggestion
== Creating the MyBatisResource to handle the requests
```

File: docs/src/main/asciidoc/mybatis.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/mybatis.adoc:10]:
```suggestion
This guide demonstrates how your Quarkus application can use link:https://mybatis.org/mybatis-3/[MyBatis] to support custom SQL, stored procedures and advanced mappings.
```

File: docs/src/main/asciidoc/mybatis.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/mybatis.adoc:43]:
```suggestion
    -Dextensions="mybatis,resteasy-jackson,jdbc-mysql"
```

**Review by gastaldi — CHANGES_REQUESTED** [review]
Found some minor typos in the doc

File: docs/src/main/asciidoc/mybatis.adoc
Comment (zhfeng) [reply docs/src/main/asciidoc/mybatis.adoc]: Done

File: docs/src/main/asciidoc/mybatis.adoc
Comment (zhfeng) [reply docs/src/main/asciidoc/mybatis.adoc]: Done

File: docs/src/main/asciidoc/mybatis.adoc
Comment (zhfeng) [reply docs/src/main/asciidoc/mybatis.adoc]: Done

**Comment by zhfeng** [comment]
@gastaldi can you take a review of these changes ?

File: docs/src/main/asciidoc/mybatis.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/mybatis.adoc:61]: What `groupId` is that?

**Comment by gastaldi** [comment]
I think this is cool. Can you start a conversation on https://groups.google.com/forum/#!forum/quarkus-dev about this extension before we merge this PR? Maybe someone could help on figuring out how to make the native build smoother for this extension

File: docs/src/main/asciidoc/mybatis.adoc
Comment (zhfeng) [reply docs/src/main/asciidoc/mybatis.adoc:61]: That should be ```io.quarkus```, I will fix it.

**Comment by zhfeng** [comment]
Thanks @gastaldi - I will start a thread on the quarkus forum.

**Comment by hantsy** [comment]
@zhfeng If possible, move this extension to your own repo as a community support extension. I do not think there are lots of developers outside of China who use MyBatis in daily work. Personally I have not used Batis for the last 10 years, never used MyBatis.

**Comment by aloubyansky** [comment]
Even if it happens to be true, btw, the decision should not depend on popularity in any specific country. As to the home for the extension, in my view, it's a question of maintainability. @zhfeng could you clarify your intentions wrt this work? Are you planning to be actively maintaining it?

**Comment by zhfeng** [comment]
I wrote a article to introduce Quarkus in Chinese two months ago and got some feedback from the developers that they want to see the MyBatis support in Quarkus which could be very helpful and attract more people to have a try.
Also there are two issue https://github.com/quarkusio/quarkus/issues/6490 and https://github.com/quarkusio/quarkus/issues/1958 in Quarkus and https://github.com/apache/camel-quarkus/issues/1384 in Camel Quarkus to show the interesting of the MyBatis extension.

@aloubyansky yeah, I'm planning to maintain this extension. The next work could be

- Add the support of XML mapping
- Support ```@Cache``` annotation in the native
- Get rid of the ```--report-unsupported-elements-at-runtime``` in the native image building

**Comment by zhfeng** [comment]
@gastaldi I just update to fix the conflict and test with ```@CacheNamespace```. I tried to take a look at the mybatis source codes deeply and it looks like they are using ```MethodHandles``` in the [MapperProxy](https://github.com/mybatis/mybatis-3/blob/master/src/main/java/org/apache/ibatis/binding/MapperProxy.java#L166) to invoke the default method of the interface dynamically. But according to the https://github.com/oracle/graal/blob/master/substratevm/LIMITATIONS.md#invokedynamic-bytecode-and-method-handles, these seem not supported by the Graal VM currently.

Also I found the https://github.com/mybatis/mybatis-3/issues/1552 which are looking for the Graal support in MyBatis. So there will be chances we can work together with the Mybatis community to support the GraalVM.

In the perspective of the Quarkus, is there anything I can do before merging this PR ?

**Comment by gastaldi** [comment]
@zhfeng thanks for the update, you've done an excellent job in this. 
I think we need to first resolve the need for ` --report-unsupported-elements-at-runtime` since that means that would require some extra work (substituting classes for example), given that none of our extensions require that.

Meanwhile if you are interested, it would be really cool if this extension ended up in https://github.com/quarkus-extension instead. WDYT @aloubyansky ?

**Comment by zhfeng** [comment]
yeah, I'm working to get rid of  ```--report-unsupported-elements-at-runtime```.

**Comment by zhfeng** [comment]
Hi @gastaldi ,

I still got some analyist parse erros during the native image builidng. And sometimes, these errors were very different in every building. Most of these are related to the ```java.lang.Class.invoke```. I'm not very sure how to fix these error in the mybatis by using some substition classes. Can you give some hints please ?

**Comment by gastaldi** [comment]
@zhfeng hey! My knowledge on writing substitution classes is based on trial and error, so I am not sure I can give some hints in that regard :)

**Comment by dhaalves** [comment]
Any update on this?

**Comment by gastaldi** [comment]
Hi @dhaalves!

We strongly recommend our community extensions to reside in the https://github.com/quarkiverse organization. This means free repository hosting (including build, CI and release publishing setup) for Quarkus extension projects contributed by the community. Plus it should integrate with our Quarkus Registry - making it visible in code.quarkus.io and related tooling (we're still working to make this happen automatically).

@zhfeng are you interested in moving this code to a `quarkiverse-mybatis` repository in the `quarkiverse` organization and become the project lead of it? I'll be more than happy to help you set up the organization and get you started ASAP. 

More details here: https://github.com/quarkiverse/quarkiverse/wiki

**Comment by zhfeng** [comment]
Thanks @gastaldi - I'm happy to move to quarkiverse.

**Comment by gastaldi** [comment]
@zhfeng fantastic! I'l create the project tomorrow morning and give you the necessary push permissions. 

Ping me.in Zulip if you have any questions.

**Comment by gastaldi** [comment]
@zhfeng Repository created in https://github.com/quarkiverse/quarkiverse-mybatis. You should have received an invitation to join the organization (You should have the Maintainer role there).

Closing this PR now

**Pull request closed without merging**
